derive/src/utils.rs

```diff
@@ -37,32 +38,99 @@ where
 	})
 }
 
-/// Look for a `#[scale(index = $int)]` attribute on a variant. If no attribute
-/// is found, fall back to the discriminant or just the variant index.
-pub fn variant_index(v: &Variant, i: usize) -> TokenStream {
-	// first look for an attribute
-	let index = find_meta_item(v.attrs.iter(), |meta| {
-		if let NestedMeta::Meta(Meta::NameValue(ref nv)) = meta {
-			if nv.path.is_ident("index") {
-				if let Lit::Int(ref v) = nv.lit {
-					let byte = v
-						.base10_parse::<u8>()
-						.expect("Internal error, index attribute must have been checked");
-					return Some(byte);
+/// Indexes used while defining variants.
+pub struct UsedIndexes {
+	/// Map from index of the variant to it's attribute or definition span
+	used_set: HashMap<u8, Span>,
+	/// We need this u8 to correctly assign indexes to variants
+	/// that are not annotated by coded(index = ?) or explicit discriminant
+	current: u8,
+}
+
+impl UsedIndexes {
+	/// Build a Set of used indexes for use with #[scale(index = $int)] attribute or
+	/// explicit  discriminant  on the variant
+	pub fn from<'a, I: Iterator<Item = &'a Variant>>(values: I) -> syn::Result<Self> {
+		let mut map: HashMap<u8, Span> = HashMap::new();
+		for v in values {
+			if let Some((index, nv)) = find_meta_item(v.attrs.iter(), |meta| {
+				if let NestedMeta::Meta(Meta::NameValue(ref nv)) = meta {
+					if nv.path.is_ident("index") {
+						if let Lit::Int(ref v) = nv.lit {
+							let byte = v
+								.base10_parse::<u8>()
+								.expect("Internal error, index attribute must have been checked");
+							return Some((byte, nv.span()));
+						}
+					}
+				}
+				None
+			}) {
+				if let Some(span) = map.insert(index, nv.span()) {
+					let mut error = syn::Error::new(nv.span(), "Duplicate variant index. qed");
+					error.combine(syn::Error::new(span, "Variant index already defined here."));
+					return Err(error)
+				}
+			} else if let Some((
+				_,
+				expr @ syn::Expr::Lit(ExprLit { lit: syn::Lit::Int(lit_int), .. }),
```

Comment:
I think this doesn't capture any discriminant, user can write this:
```rust
#[derive(Encode, Decode)]
enum A {
	A = 3 + 4,
}
```

I think it can be ok to constraint our implementation, but it should be done in a safe way, maybe requiring a new attribute which specifies legacy implementation or new implementation.

Reply:
updated with to disallow this pattern